lib/charms/legend_database_manager/v0/database_manager.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2021 Canonical
+# See LICENSE file for licensing details.
+
+import json
+
+from ops import framework
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "431732f8afb641a3a5a38e5c5d01ee11"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 0
+
+REQUIRED_LEGEND_DATABASE_CREDENTIALS = [
+    "username", "password", "database", "uri"]
+
+
+class LegendDatabaseConsumer(framework.Object):
+    def __init__(self, charm, relation_name="legend-db"):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+
+    def database_connection_from_mongodb(
+            self, mongodb_consumer, mongodb_relation_id):
+        """Returns a dict with Mongo connection info for Legend components
+        just like `get_legend_database_creds`.
+
+        Args:
+            mongodb_consumer:  charms.mongodb_k8s.v0.mongodb.MongoConsumer
+            mongodb_relation_id: ID of the MongoDB relation.
+
+        Returns:
+            Dictionary with the following structure:
+            {
+                "uri": "<replica set URI (with user/pass, no DB name)>",
+                "username": "<username>",
+                "password": "<password>",
+                "database": "<database name>"
+            }
+        """
+        creds = mongodb_consumer.credentials(rel_id=mongodb_relation_id)
+        if not creds:
+            return {}
+
+        databases = mongodb_consumer.databases(rel_id=mongodb_relation_id)
+        if not databases:
+            return {}
+
+        uri = creds['replica_set_uri']
+        # NOTE: we remove the trailing database from the URI:
+        split_uri = [
+            elem for elem in uri.split('/')[:-1]
+            # NOTE: filter any empty strings resulting from double-slashes:
+            if elem]
+        # NOTE: schema prefix needs two slashes added back:
+        uri = "%s//%s" % (
+            split_uri[0], "/".join(split_uri[1:]))
+
+        return {
+            "uri": uri,
+            "username": creds['username'],
+            "password": creds['password'],
+            "database": databases[0]}
+
+    def _validate_legend_database_credentials(self, creds):
+        """Returns True/False depending on whether the provided Legend
+        database credentials dict contains all the required keys."""
+        if any([not creds.get(k)
+                for k in REQUIRED_LEGEND_DATABASE_CREDENTIALS]):
+            return False
+        return True
+
+    def get_legend_database_creds(self, relation_id):
```

Comment:
This function looks like it returns the same information as `database_connection_from_mongodb()` so I am not sure why have both.

Reply:
This method returns the information from the LDBM-Legend relation and is meant to be called on the receiving end of that relation in the SDLC/Engine/Studio.

As mentioned, the other method is only meant to be called by the LDBM in the context of its relation to Mongo.